### PR TITLE
Cache responses of xy_to_pir_API_single

### PR DESCRIPTION
**Pull request by Stannislav**
This is currently the bottleneck when running `atldld dataset preview` and `atldld dataset download-faithful` with the same parameters again and again. This is because for every section image the four corners in ref space have to be queried.

Testing: run `atldld dataset preview {479, 75492803, ...}` multiple times.

### CODE REVIEW

**Comment by jankrepl** [comment]
IMO if we merge #71 this will be redundant. However, I think we can definitely think of caching the metadata (similarly as we do with the images).

**Comment by Stannislav** [comment]
> IMO if we merge #71 this will be redundant. However, I think we can definitely think of caching the metadata (similarly as we do with the images).

Yes, that's true. What are we going to do with `xy_to_pir_API_single`?

As long as  `xy_to_pir` is not completely integrated everywhere it's still useful to cache though :)

**Comment by jankrepl** [comment]
> > IMO if we merge #71 this will be redundant. However, I think we can definitely think of caching the metadata (similarly as we do with the images).
> 
> Yes, that's true. What are we going to do with `xy_to_pir_API_single`?
> 
> As long as `xy_to_pir` is not completely integrated everywhere it's still useful to cache though :)

I would keep `xy_to_pir_API_single` in the codebase, since it was actually used to create the test data for that PR. And maybe in the future we will want to create more test data (we might discover some edge cases that the current test data does not test).

Sure, we can merge this PR for now:)

File: tests/test_utils.py [hunk 1]
```diff
@@ -117,6 +120,64 @@ def xy_to_pir(x, y, _image_id):
...
 
+class TestXyToPirApiSingle:
+    @pytest.mark.internet
+    @pytest.mark.parametrize("image_id", EXISTING_IMAGE_IDS)
+    def test_fetching_online_no_mocking(self, image_id):
+        x, y = np.random.randint(0, 10000, 2)
+        p, i, r = xy_to_pir_API_single(x, y, image_id=image_id)
+        assert np.isfinite(p)
+        assert np.isfinite(i)
+        assert np.isfinite(r)
```
Comment (EmilieDel) [inline tests/test_utils.py:131]: Why are we not checking the exact values if we do a real requests anyway ?

Comment (EmilieDel) [inline tests/test_utils.py:131]: Why are we not checking the exact values if we do a real requests anyway ?

**Review by EmilieDel — APPROVED** [review]
Good idea 👍

File: tests/test_utils.py (hunk 1 above)
Comment (Stannislav) [reply tests/test_utils.py:131]: Ah but `x` and `y` are always random, so to get the real PIR value one would be forced to kind of re-implement the logic of the actual function, which I didn't really like to do...

By the way, this `np.isfinite` check was there before but in a slightly different setting.

I think an alternative could be to hard-code `x` and `y` for each of the dataset IDs, but I don't know. What do you think?

Comment (EmilieDel) [reply tests/test_utils.py:131]: I see your point! I am just wondering if this test could really detect some bugs by simply checking if the number are finite 🤔

Comment (Stannislav) [reply tests/test_utils.py:131]: I think you're right, these tests are probably not telling us much.

What about this kind of test?
```python
    @pytest.mark.internet
    @pytest.mark.parametrize("image_id", EXISTING_IMAGE_IDS)
    def test_fetching_online_no_mocking(self, image_id):
        x, y = np.random.randint(0, 10000, 2)
        pir = xy_to_pir_API_single(x, y, image_id=image_id)
        assert len(pir) == 3
        assert (isinstance(coordinate, float) for coordinate in pir)
```
But it actually does pretty much the same thing as before - the previous test would also fail if the length isn't 3 or if the coordinates are not numbers.

We can also just skip the asserts and just write
```python
xy_to_pir_API_single(x, y, image_id=image_id)
```
which would just test if the online queries work in principle and don't raise any errors. The returned values and types are then covered by other (offline) tests.

What do you think?

Comment (EmilieDel) [reply tests/test_utils.py:131]: You are right! 

I agree, It is already good to test that the URL is still valid and does not raise any error + that we have 3 numbers as output! We can keep as is! Thanks for the reply @Stannislav!

**Review by jankrepl — APPROVED** [review]
